apps/prairielearn/assets/scripts/instructorAssessmentAccessClient.ts

```diff
@@ -0,0 +1,225 @@
+    refreshTable();
+  }
+
+  on('click', '.editButton', (e) => {
+    const row = parseInt((e.target as HTMLElement).closest('button')?.dataset.row ?? '0');
+    const accessRule = accessRulesData[row];
+    accessRule.formatted_start_date = adjustedDate(accessRule.start_date)
+      .toISOString()
+      .slice(0, 19);
+    accessRule.formatted_end_date = adjustedDate(accessRule.end_date).toISOString().slice(0, 19);
+    if (!timeZoneName) {
+      throw new Error('Course instance time zone is not available.');
+    }
+    morphdom(
+      editAccessRuleModal as Node,
+      EditAccessRuleModal({ accessRule, i: row, timeZoneName }).toString(),
+    );
+    $('#editAccessRuleModal').modal('show');
+  });
+
+  on('click', '#updateAccessRuleButton', (e) => {
+    const row = parseInt((e.target as HTMLElement).closest('button')?.dataset.row ?? '0');
+    const form = (e.target as HTMLElement).closest('form');
+    if (!form) return;
+    handleUpdateAccessRule({ form, row });
+  });
+
+  function swapRows(row: any, targetRow: any) {
+    const accessRule = accessRulesData[row];
+    accessRulesData[row] = accessRulesData[targetRow];
+    accessRulesData[targetRow] = accessRule;
+    refreshTable();
+  }
+
+  on('click', '.up-arrow-button', (e) => {
+    const row = parseInt((e.target as HTMLElement).closest('button')?.dataset.row ?? '0');
+    if (!row) return;
+    swapRows(row, row - 1);
+  });
+
+  on('click', '.down-arrow-button', (e) => {
+    const row = parseInt((e.target as HTMLElement).closest('button')?.dataset.row ?? '0');
+    if (row === accessRulesData.length - 1) return;
+    swapRows(row, row + 1);
+  });
+
+  on('click', '#addRuleButton', () => {
+    const formatted_start_date = adjustedDate(new Date()).toISOString().slice(0, 19);
+    const formatted_end_date = adjustedDate(new Date()).toISOString().slice(0, 19);
+    if (!timeZoneName) {
+      throw new Error('Course instance time zone is not available.');
+    }
+    morphdom(
+      editAccessRuleModal as Node,
+      EditAccessRuleModal({
+        accessRule: {
+          mode: '',
+          uids: '',
+          start_date: new Date().toISOString().slice(0, 19),
+          end_date: new Date().toISOString().slice(0, 19),
+          active: 'True',
+          credit: '100',
+          time_limit: '50',
+          password: '',
+          formatted_start_date,
+          formatted_end_date,
+          exam_uuid: null,
+          pt_course_id: null,
+          pt_course_name: null,
+          pt_exam_id: null,
+          pt_exam_name: null,
+        },
+        i: accessRulesData.length,
+        addAccessRule: true,
+        timeZoneName,
+      }).toString(),
+    );
+    $('#editAccessRuleModal').modal('show');
+  });
+
+  on('click', '.deleteButton', (e) => {
+    const row = parseInt((e.target as HTMLElement).closest('button')?.dataset.row ?? '0');
+    morphdom(deleteAccessRuleModal as Node, DeleteConfirmationModal().toString());
```

Comment:
Is morphdom actually necessary here? AFAICT the modal is actually static and doesn't ever need to be changed on the client.

Reply:
No, it is not. I have updated this to use `templateFromAttributes` in [1dfef07](https://github.com/PrairieLearn/PrairieLearn/pull/9934/commits/1dfef07527ed0296ff3cdfa7c58d85fce2304b97).